### DISCUSSION TITLE
How to prevent redirect - "Stripe Elements & PaymentIntents (no redirect)" - I can't avoid redirect

### DISCUSSION THREAD

**Discussion by drscottlobo** [vercel/next.js#33846] [Help]
Hello, 

I'm having an issue with the Elements version of the example in the folder "with-stripe-typescript". I'm trying to keep the form on the client side while confirming the paymentIntent. 

I'm using the following pages from the docs per the instructions: 

**Stripe Elements**
Custom Amount Donation with Stripe Elements & PaymentIntents **(no redirect)**:
Frontend: pages/donate-with-elements.tsx
Backend: pages/api/payment_intents/

The live demo found [here](https://nextjs-typescript-react-stripe-js.vercel.app/donate-with-elements) does not redirect when a test payment is made. 

However, every time I try to make the same functionality work with my local environment, it ALWAYS redirects on a successful transaction, and I believe it is a result of the required "return_url" param calling a redirect in the following code here: [next.js/examples/with-stripe-typescript/components/ElementsForm.tsx:](https://github.com/stripe-samples/nextjs-typescript-react-stripe-js/blob/master/components/ElementsForm.tsx)

  ```
// Use your card Element with other Stripe.js APIs
  const { error } = await stripe!.confirmPayment({
      elements,
      confirmParams: {
        return_url: 'http://localhost:3000/donate-with-elements', //// HERE?
        payment_method_data: {
          billing_details: {
            name: input.cardholderName,
          },
        },
      },
    })

    if (error) {
      setPayment({ status: 'error' })
      setErrorMessage(error.message ?? 'An unknown error occurred')
    } else if (paymentIntent) {
      setPayment(paymentIntent)
    }
  }
```
Is the demo not using this code? Is some other API being used and therefore not having the issue with Stripe redirecting to my donate-with-elements page? Or am I just missing something somehow? 

I'm also very confused by the comment "Use your card Element with other Stripe.js APIs"? Is the function that follows the comment not not used in the live demo? The comment seems to suggest this is an example of an alternative? 

Either way, I would much appreciate a way to confirm the payment without the redirect being caused by stripe!.confirmPayment(). There must be a way, since the live demo Elements version does not redirect.

Thanks

**Comment by nickhalldev** [comment]
I struggled to find as well. Looks like in confirmPayments you can make it so it only redirects if it is required. 
`
.confirmPayment({
        elements,
        confirmParams: {
          return_url: 'https://localhost:3000/',
        },
        redirect: 'if_required',
      })
`

[https://stripe.com/docs/js/payment_intents/confirm_payment?type=pii#confirm_payment_intent-options-redirect](Stripe doc)

**Reply by nabilfreeman** [reply]
Doesn't this mean that there are still some scenarios where it will redirect? Can't we disable redirecting completely?

**Reply by muhamedh** [reply]
My understanding is at certain time when a payment requires 3DS it needs to know to redirect to your page after the user approves the payment

**Reply by nabilfreeman** [reply]
Yes, I recently had a call with a Solution Architect at Stripe and there is no way to 100% prevent redirects when using `confirmPayment`. 3D Secure will open in a modal and not redirect, but for other payment methods the only way to prevent/avoiding coding it is to disable non-card payment methods.

**Reply by muhamedh** [reply]
Is there a different option to `confirmPayment` ?

**Reply by michaelbutler** [reply]
So, I have done some thought about this and if you truly don't want to ever have redirects -- because maybe you want to maintain the client-side state which gets killed by redirects -- there are a couple of ideas with varying pros & cons:

- Set up a more robust client-side state system so that entered data is stored in LocalStorage, for example. This way, when a redirect does occur, when the user comes back, all the data entered is still there. 
- Create your own JavaScript pop-up window to handle Stripe logic. The basis of this is the `window.open(URL)` function. Problem is, most browsers disable these kind of pop-ups by default, and they might have to click an Allow pop-up button which isn't a great experience. Supposedly if the window.open call is in an onClick handler, it has better chances of working.
- Similar (and perhaps more reliable) to the previous, is instead of using `window.open`, you can just have an anchor tag like `<a href="/handle-payments" target="_blank">Click me</a>`.  Upon clicking, the user will get sent to a new browser tab. At the end of the process you can use `window.close()` (which actually might be prevented with security restrictions). Or you can instruct the user to close the window to continue.

In any of these cases, you'll probably need to maintain some server side logic that the client side can communicate with asynchronously about the user's payment & checkout state, perhaps through periodic polling as your server will be getting callback hooks directly from Stripe's servers.
